Scheduler does not check permissions on server_dyn_res script

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
PBS does not check for permissions of server/mom dynamic resource script. There can be a case that the permissions of this file are open for 'write' to others and group.

#### Affected Platform(s)
All

#### Cause / Analysis / Design
No permission checks were performed.


#### Solution Description
Added permission checks and also added a check that the user owning the script should have a uid <10 or be an admin (on windows)

#### Testing logs/output
[test_momdynres.txt](https://github.com/PBSPro/pbspro/files/2631339/test_momdynres.txt)
[test_svrdynres.txt](https://github.com/PBSPro/pbspro/files/2631340/test_svrdynres.txt)
[test_smoketest.txt](https://github.com/PBSPro/pbspro/files/2631342/test_smoketest.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__



#### NOTE
This pull request is copy of "https://github.com/PBSPro/pbspro/pull/905" which was failing to load and possibly causing Travis/appveyor to not run

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11093,21 +11103,19 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
-            self.du.chmod(self.hostname, path=res_file, mode=perm, sudo=True)
+            with open(res_file) as f:
+                script_body = f.readlines()
+                self.du.chmod(self.hostname, path=res_file, mode=perm,
```
I don't think this chmod() has to be in the with statement.  The with statement should be the code that deals with the open file.  This happened after the close before, so shouldn't it happen after the with statement now?

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -13497,21 +13505,19 @@ def add_mom_dyn_res(self, custom_resource, script_body=None,
-            self.du.chmod(self.hostname, path=res_file, mode=perm, sudo=True)
+            with open(res_file) as f:
+                script_body = f.readlines()
+                self.du.chmod(self.hostname, path=res_file, mode=perm,
```
Same here, I think this should be outside the with statement

arungrover commented:
@bhroam and @agrawalravi90 thanks for reviewing this pull request. @agrawalravi90 You must have noticed that this is a new pull request. I had to do that because the previous pull was not loading most of the time. Another problem was that the previous pull was not even triggering CI (travis or appveyor) because it had conflicts with the master. I have not seen github doing this earlier, when I rebased my branch, it started working.
Now about new changes - I have reverted create_temp_file to what it was. It is because Travis started failing while running "test_staging" smoke test. This was because travis runs tests as root and this test was trying to create a temp file as TEST_USER. Now new create_temp_file creates a file in user's (who is running benchpress) home and then copy it to "asuser's" home directory. This copy was failing because TEST_USER did not have read access to root's home directory.
At this point, I have given up on changing create_temp_file. Instead, I am not doing a copy, chown and chmod in add_server/mom_dyn_res() methods. I have also reverted the selftest I added in the last commit.

I've attached the test result again. please have a look.
[test_momdynres.txt](https://github.com/PBSPro/pbspro/files/2917594/test_momdynres.txt)
[test_svrdynres.txt](https://github.com/PBSPro/pbspro/files/2917597/test_svrdynres.txt)

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11110,13 +11110,15 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
+                             preserve_permission=False)
+            self.du.chown(path=res_file, uid=0, gid=0, sudo=True)
+            self.du.chmod(self.hostname, path=res_file, mode=perm, sudo=True)
             self.dyn_created_files.append(res_file)
```
As per our offline discussion, I thought you were going to revert this since @latha-subramanian is implementing this for all files created by create_temp_file()

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11110,13 +11110,15 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
+            self.du.chmod(self.hostname, path=res_file, mode=perm, sudo=True)
             self.dyn_created_files.append(res_file)
-
         self.server_dyn_res.append(res_file)
```
If you do end up keeping this, won't you add the file twice (once the line above and once here)?

arungrover replied:
Well this is not exactly the file create_temp_file creates. This one is a copy that we make in PBS_HOME. That is why I kept this code as is.

arungrover replied:
I think I misplaced my comment here. It was meant for your first comment. You are right that the file is getting added twice. I'm thinking to remove the existing implementation where we delete server_dyn_res files in revert_to_defaults. This is because the last test will leave its file in the system.

arungrover commented:
I found some more bugs where I was not passing hostname to dshutil methods properly. I'll push that change along with codacy's suggested changes.

arungrover commented:
I'm rebasing the branch to get the latest IFL changes.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11079,25 +11090,45 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
+                          sudo=True)
+            self.du.chmod(hostname=host, path=res_file, mode=perm, sudo=True)
+            if host is None:
+                self.dyn_created_files.append(res_file)
```
i think DshUtuls's rm() can delete files on remote hosts as well

arungrover replied:
You are right, but I'm not storing hostname along with the files. Although that can be done but since it is such a remote case for our tests, I just left it. Also, since the file is created inside PBS_HOME it gets cleaned up when the PBS is uninstalled in out testbed machines.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11079,25 +11090,45 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
+        :param prefix: the file name will begin with this prefix
+        :type prefix: str
+        :param suffix: the file name will end with this suffix
+        :type suffix: str
```
Please add a :return to describe what this function returns since it seems to return the name of the file created.

agrawalravi90 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -13451,6 +13478,61 @@ def action(self, act, script):
+            res_file = os.path.join(dirname, tmp_file.split(os.path.sep)[-1])
+            self.du.run_copy(host, tmp_file, res_file, sudo=True,
+                             preserve_permission=False)
+            self.du.chown(hostname=host, path=res_file, uid=0, gid=0,
```
same comment here, can you please mention in the docstring that the file created will be root owned.

agrawalravi90 commented on test/fw/ptl/utils/pbs_testsuite.py:
```diff
@@ -1471,6 +1471,15 @@ def tearDown(self):
         self.stop_proc_monitor()
+
+        for server in self.servers.values():
+            server.cleanup_files()
```
shouldn't it be scheduler.cleanup_files() ?

arungrover replied:
Well, I added cleanup_files() in the PBSService base class. I thought in future if there is a need for the server to clean up files, this function can provide that functionality. As of now, the server does nothing when cleanup_files() is called.

agrawalravi90 replied:
sorry, i didn't see the scheduler.cleanup_files() line below. That brings another thought: do you need to call scheduler.cleanup_files() for all scheduler objects (in case of multi-sched) ?

arungrover commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -11079,25 +11090,45 @@ def add_server_dyn_res(self, custom_resource, script_body=None,
+            res_file = os.path.join(dirname, tmp_file.split(os.path.sep)[-1])
+            self.du.run_copy(host, tmp_file, res_file, sudo=True,
+                             preserve_permission=False)
+            self.du.chown(hostname=host, path=res_file, uid=0, gid=0,
```
You are right, they all go together. I will add a comment about file being owned by root.

arungrover commented:
Thanks @agrawalravi90 for reviewing the change. I've addressed your review comments.

agrawalravi90 commented:
Looks good Arun! Thanks for being patient

arungrover commented:
Thanks @agrawalravi90 and @bhroam for reviewing it.
I've squashed and rebased. I'll let you know once CI passes.

bhroam commented:
I merged in the PR.
